aten/src/ATen/cudnn/Descriptors.h

```diff
@@ -193,12 +193,7 @@ struct AT_CUDA_API ConvolutionDescriptor
     if (dataType == CUDNN_DATA_HALF) mathType = CUDNN_DATA_FLOAT;
     AT_CUDNN_CHECK(cudnnSetConvolutionNdDescriptor(mut_desc(), dim, pad, stride, upscale,
                                           CUDNN_CROSS_CORRELATION, mathType));
-#if CUDNN_VERSION >= 7000
     AT_CUDNN_CHECK(cudnnSetConvolutionGroupCount(mut_desc(), groups));
-    AT_CUDNN_CHECK(cudnnSetConvolutionMathType(mut_desc(), CUDNN_DEFAULT_MATH));
```

Comment:
I guess, what I expected, was to see mathType added as an argument to this method, and then a set to cudnnSetConvolutionMathType here (instead of twiddling it in multiple locations further down this diff, where you could easily forget to put one site in.)

Reply:
Yeah, so I finally actually reviewed the patch, and I understand why I am confused.

The invariant that I want to see out of this code, is that the default convolution descriptor (when I call set() on it) has "good" settings. By deleting these lines, we end up with a descriptor that is default everything, even though we know that in the absence of benchmarking, we do subsequently toggle `CUDNN_TENSOR_OP_MATH` on for half inputs. So, to me, it seems like it would be better to set this here, and then delete the code that does the equivalent modification at https://github.com/pytorch/pytorch/blob/fc574b53f42a2d0fbaf0228faa65689fb4a3fbb9/aten/src/ATen/native/cudnn/Conv.cpp#L706 It also saves you from having to insert this code in multiple code paths.

What do you think?

Reply:
That's totally doable. I was debating between whether to call cudnnSetConvolutionMathType multiple times vs putting those extra lines regarding default mathtype and algo. But since cudnnSetConvolutionMathType is cheap, let's bring those lines back and get rid of the multiple code paths regarding mathType.

Reply:
Please see the latest changes (https://github.com/pytorch/pytorch/pull/15881/commits/014b7b9e66581da8f1478c53817c765390eaeea5). I think it addresses both of our confusion?

Reply:
No, I am still confused. In the latest patch, you are still not setting `cudnnSetConvolutionMathType` inside `ConvolutionDescriptor`, instead, you're returning `cudnnMathType_t` via pointer argument, which is weird and pretty out of line with how the rest of the descriptors work.

Also, I don't understand why it is necessary to pre-fill `fwdAlgPerf.mathType` with a math type, in your code. The invariant I would expect is `chooseAlgorithm` always leaves `fwdAlgPerf` in a valid state. Is the problem this code?

```
  if (args.params.deterministic && !benchmark) {
    *algo = search::DEFAULT_ALGO;
    return;
  }
```

I don't know, but I could imagine that the `DEFAULT_ALGO` here isn't half aware and so you get the wrong math type in that case? Sure, but in that case, I would expect up the fixup for half case to occur here.

Reply:
You might have already reiterated these above, but let me think out loud and let's start with the case that `cudnnSetConvolutionMathType` must be called after `chooseAlgorithm` (because we need to update the cdesc mathType with the resultant mathType of `chooseAlgorithm`, where `chooseAlgorithm` has three paths - cudnnFind, cudnnGet and default). Because of this case, I cannot call cudnnSetConvolutionMathType in the descriptor's setter, because the `cudnnSetConvolutionMathType` called after `chooseAlgorithm` would overwrite the effect of the one in the descriptor's setter, and it would overwrite it with garbage for the default path if I don't have the following in `chooseAlgorithm`.
```
if (dataType == CUDNN_DATA_HALF) {
  algoPerf->mathType = CUDNN_TENSOR_OP_MATH;
} else {
  algoPerf->mathType = CUDNN_DEFAULT_MATH;
}
```
So that's why I have only two ways for resolving this: 
- either sprinkle that above if-else everywhere in `chooseAlgorithm` to make sure we don't set cdesc mathType to garbage when calling `cudnnSetConvolutionMathType` after `chooseAlgorithm`, for the default path.
- or pre-fill `fwdAlgPerf.mathType` in the cdesc setter so that the mathType in `cudnnConvolutionFwdAlgoPerf_t` is in good state for the default path, and we can avoid those if-else's.

All these so that I could remove traces of `cudnnConvolutionFwdAlgo_t` and use the mathType and algo members of `cudnnConvolutionFwdAlgoPerf_t`
```
typedef struct {
    cudnnConvolutionFwdAlgo_t algo;
    cudnnStatus_t status;
    float time;
    size_t memory;
    cudnnDeterminism_t determinism;
    cudnnMathType_t mathType;
    int reserved[3];
} cudnnConvolutionFwdAlgoPerf_t;
```
Should I have just used `cudnnConvolutionFwdAlgo_t algo` and `cudnnMathType_t mathType` separately then?

Reply:
> cannot call cudnnSetConvolutionMathType in the descriptor's setter, because the cudnnSetConvolutionMathType called after chooseAlgorithm would overwrite the effect of the one in the descriptor's setter, and it would overwrite it with garbage for the default path if I don't have the following in chooseAlgorithm.

No, I don't agree with this. Look at the code here:

```
  args.cdesc.set(dataType, input.dim() - 2, args.params.padding, args.params.stride, args.params.dilation, args.params.groups);

  // TODO: when we do legacy group convolution support, we'll repeatedly
  // reinitialize the workspace for each convolution we do.  This is
  // wasteful; we'd rather reuse the workspace.  OTOH, legacy group
  // convolution support is already pretty slow, so this might not
  // matter.  (This applies to raw_cudnn_convolution_backward_input as well.)
  cudnnConvolutionFwdAlgo_t fwdAlg;
  Workspace workspace = chooseAlgorithm(args, benchmark, &fwdAlg);
```

The `set` method on descriptor is clearly called before `chooseAlgorithm`.  I additionally audited every other occurrence of `set(` in native/cudnn/Conv.cpp to convince myself that ConvolutionDescriptor::set is always called before chooseAlgorithm in all codepaths.

I certainly agree with you that in the event that benchmarking occurs, we may need to subsequently call cudnnSetConvolutionMathType after benchmarking, to update it (thus leading to an extra, technically unnecessary, setting of math type. For interest of code clarity, I am recommending that you do this extra work.)

I didn't respond to your paragraphs before this claim. Let me know if I should.

Reply:
Ok. I prepared a diff for you based on your most recent reply (https://github.com/syed-ahmed/pytorch/commit/0499daa57f5af4da6c4464e412cb738cd1dba310) but I'm embarrassed to say that I've discovered a cudnn bug! :'(

The bug is, if I call cudnnSetConvolutionMathType in the setter of the descriptor (which sets the mathType to CUDNN_TENSOR_OP when fp16), cudnnGet*_v7 returns algo1 with CUDNN_TENSOR_OP math type instead of not caring about what was set by cudnnSetConvolutionMathType before it (and returning algo1 with CUDNN_DEFAULT_MATH which is performant)! 

Any suggestions on how can I move forward?

Reply:
Interesting. Sorry about sending you on this wild goose chase!

In that case, I think something similar to an earlier version of your patch would be preferred (before you started changing it due to my comments). Here's the new justification:

1. The job of `ConvolutionDescriptor::set` is to setup a convolution descriptor so that algorithm search works for it. Since setting math type to `CUDNN_TENSOR_OP` prevents cuDNN from considering non-tensor-op algorithms, that means it's correct for it to unconditionally set `CUDNN_DEFAULT_MATH `
2. The job of `findAlgorithm` is to update ConvolutionDescriptor with correct, performant parameters, **even** if we never actually call the cudnn benchmarking function. And as the code is structured right now, you'll have to write the code to adjust the math setting in multiple places. I guess it's not worth fixing now.

So... maybe just reverting to https://github.com/pytorch/pytorch/pull/15881/commits/e5038f9d1186dfd1ab1a1a0cf240d3d996af768f and adding some comments from our investigation is sufficient? Once again, sorry about the wild goose chase!

Reply:
Oh no no. It's all good. Your points are legit. I'll put the comments in for now and hopefully we can address this again in the future.